### PR TITLE
8338658: New Object to ObjectMonitor mapping: s390x implementation

### PR DESCRIPTION
**Pull request by offamitkumar**
s390x implementation of [JDK-8315884](https://bugs.openjdk.org/browse/JDK-8315884) New Object to ObjectMonitor mapping; 

Testing: 
- tier1-test (fastdebug)
- tier1-test with UseObjectMonitorTable (fastdebug)
- tier1-test with UseObjectMonitorTable (release)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8338658](https://bugs.openjdk.org/browse/JDK-8338658): New Object to ObjectMonitor mapping: s390x implementation (**Sub-task** - P4)

### Reviewers
 * [Lutz Schmidt](https://openjdk.org/census#lucy) (@RealLucy - **Reviewer**)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20740/head:pull/20740` \
`$ git checkout pull/20740`

Update a local copy of the PR: \
`$ git checkout pull/20740` \
`$ git pull https://git.openjdk.org/jdk.git pull/20740/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20740`

View PR using the GUI difftool: \
`$ git pr show -t 20740`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20740.diff">https://git.openjdk.org/jdk/pull/20740.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20740#issuecomment-2314231439)

### CODE REVIEW

**Review by RealLucy — CHANGES_REQUESTED** [review]
I made a few comments you may want to consider.

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp [hunk 1]
```diff
@@ -6218,33 +6229,78 @@ void MacroAssembler::compiler_fast_lock_lightweight_object(Register obj, Registe
...
+      }
+
+      NearLabel loop;
+      // Search for obj in cache
+
+      bind(loop);
+
+      // check for match.
+      z_cg(obj, Address(tmp1));
+      z_bre(monitor_found);
```
[the remark this answers is not in the document]
Comment (offamitkumar) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:6255]: I couldn't find answer for that. Maybe @xmas92 can tell us about that.

[the remark this answers is not in the document]
Comment (xmas92) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:6265]: There are 8 cache entries, and a null sentinel at the end. All entries can be null. 

So the answer is no we can not be sure about that as one, two or three of the first three entries may be null. But I am not sure what the reason is for this question.

The non-empty/non-null entries always comes first, followed by the null entries if any, followed by a null sentinel. The unrolled entries do not check for null, only for a match. The loop will check the rest of the entries and go to a slow path when a null entry (or the null sentinel) is encountered.

[the remark this answers is not in the document]
Comment (RealLucy) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:6265]: Thank you , @xmas92 !
With that, the code looks correct. The unrolled iterations may find a match. All other cases are safe, no out-of-bounds. The loop iterations check for null and are thus safe as well.

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp [hunk 2]
```diff
@@ -6019,6 +6019,16 @@ void MacroAssembler::lightweight_lock(Register obj, Register temp1, Register tem
...
 
+  if (UseObjectMonitorTable) {
+    // Clear cache in case fast locking succeeds.
+    const Address om_cache_addr = Address(basic_lock, BasicObjectLock::lock_offset() + in_ByteSize((BasicLock::object_monitor_cache_offset_in_bytes())));
+    if (om_cache_addr.is_disp12()) {
+      z_mvghi(om_cache_addr, 0);
+    } else {
+      z_lghi(top, 0); // tmp1 is free at this point
+      z_stg(top, om_cache_addr);
+    }
```
Comment (offamitkumar) [inline src/hotspot/cpu/s390/macroAssembler_s390.cpp:6030]: @RealLucy should I remove the else Part ? I tested on tier1 and it is not being executed, So maybe safe to remove ?

Comment (RealLucy) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:6030]: I am often undecided myself. The code as it is now is correct for all displacements. If you omit the else part, you introduce a hidden dependency on the layout of BasicObjectLock and BasicLock. On the other hand, how likely is it that anybody will fundamentally change the layout and thus break the disp12 requirement? Compact or generally valid? **Your choice.**

Comment (offamitkumar) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:6030]: I guess let's keep it. I mean even if there is need to change the layout, then we have to remove `z_mvghi` and switch back to this implementation again. So maybe better we keep it here and hope for the best ?

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheRealMDoerr) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:6265]: @xmas92: Can the 1st entry be 0 and the 2nd one contain garbage which matches the object by chance?

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp [hunk 3]
```diff
@@ -6019,6 +6019,16 @@ void MacroAssembler::lightweight_lock(Register obj, Register temp1, Register tem
   // instruction emitted as it is part of C1's null check semantics.
   z_lg(mark, Address(obj, mark_offset));
 
+  if (UseObjectMonitorTable) {
+    // Clear cache in case fast locking succeeds.
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/s390/macroAssembler_s390.cpp:6023]: @xmas92: This comment sounds like it should only be done if fast locking succeeds. Why are we doing it regardless of that?

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (xmas92) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:6265]: No, the oops in the cache are either valid entries or null. They may be stale, as in the monitor has been deflated, but that will cause the slow path to be taken and the cache entry is replaced inside the runtime call.

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp (hunk 2 above)
Comment (TheRealMDoerr) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:6030]: `BasicObjectLock` is a very small data structure. 12 bit offsets are more than enough. The actual offsets are 0 :-)

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp [hunk 4]
```diff
@@ -6152,6 +6157,12 @@ void MacroAssembler::compiler_fast_lock_lightweight_object(Register obj, Registe
   // Finish fast lock unsuccessfully. MUST branch to with flag == EQ
   NearLabel slow_path;
 
+  if (UseObjectMonitorTable) {
+    // Clear cache in case fast locking succeeds.
+    z_lghi(tmp1, 0); // tmp1 is free at this point
+    z_stg(tmp1, Address(box, BasicLock::object_monitor_cache_offset_in_bytes()));
+  }
+
```
[the remark this answers is not in the document]
Comment (TheRealMDoerr) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:6175]: +1

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheRealMDoerr) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:6265]: Ok, the owner won't be 0 in this case, the cmpxchg attempt will fail and the slow path will handle it correctly. Thanks for the explanation!

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp (hunk 3 above)
Comment (TheRealMDoerr) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:6023]: Thanks for explaining! I'm ok with it and I have implemented it the same way for PPC64: https://github.com/openjdk/jdk/pull/20922

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp (hunk 2 above)
Comment (offamitkumar) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp]: I was only looking for extra safety. But I reverted it. Thanks for the suggestion.

**Comment by offamitkumar** [comment]
@TheRealMDoerr @RealLucy can I get further reviews on this  :-)

**Review by RealLucy — APPROVED** [review]
Changes are looking good now.

**Review by TheRealMDoerr — APPROVED** [review]

**Comment by offamitkumar** [comment]
thanks for the reviews; 

/integrate

**Pull request closed without merging**